test/cqlpy/test_vector_similarity.py

```diff
@@ -0,0 +1,363 @@
+        for row in result:
+            assert row[1] is None
+        with pytest.raises(InvalidRequest, match="Cannot infer type of argument"):
+            cql.execute(f"SELECT pk, similarity_{similarity_function}(null, null) FROM {table}")
+
+
+@pytest.mark.parametrize("similarity_function", ["cosine", "euclidean", "dot_product"])
+def test_vector_similarity_allow_both_vector_columns_and_literals_as_arguments(cql, test_keyspace, similarity_function):
+    schema = 'pk int primary key, v1 vector<float, 3>, v2 vector<float, 3>'
+    with new_test_table(cql, test_keyspace, schema) as table:
+        cql.execute(f"INSERT INTO {table} (pk, v1, v2) VALUES (1, [1.0, 2.0, 3.0], [4.0, 5.0, 6.0])")
+        result = cql.execute(f"SELECT pk, similarity_{similarity_function}([1.1, 1.2, 20.25], [1.8, 0.5, 20.03]) FROM {table}")
+        for row in result:
+            assert_close(row[1], compute_similarity(similarity_function, [1.1, 1.2, 20.25], [1.8, 0.5, 20.03]))
+        result = cql.execute(f"SELECT pk, v1, v2, similarity_{similarity_function}(v1, v2) FROM {table}")
+        for row in result:
+            assert_close(row[3], compute_similarity(similarity_function, row.v1, row.v2))
+        result = cql.execute(f"SELECT pk, v1, similarity_{similarity_function}(v1, [1.8, 0.5, 20.03]) FROM {table}")
+        for row in result:
+            assert_close(row[2], compute_similarity(similarity_function, row.v1, [1.8, 0.5, 20.03]))
+        result = cql.execute(f"SELECT pk, v2, similarity_{similarity_function}([1.1, 1.2, 20.25], v2) FROM {table}")
+        for row in result:
+            assert_close(row[2], compute_similarity(similarity_function, [1.1, 1.2, 20.25], row.v2))
+
+
+@pytest.mark.parametrize("similarity_function", ["cosine", "euclidean", "dot_product"])
+def test_vector_similarity_fails_on_vector_of_different_size(cql, test_keyspace, similarity_function):
+    schema = 'pk int primary key, v vector<float, 3>'
+    with new_test_table(cql, test_keyspace, schema) as table:
+        cql.execute(f"INSERT INTO {table} (pk, v) VALUES (1, [1.0, 2.0, 3.0])")
+        with pytest.raises(InvalidRequest, match="All arguments must have the same vector dimensions"):
+            cql.execute(f"SELECT pk, similarity_{similarity_function}(v, [1.1, 1.2]) FROM {table}")
+        with pytest.raises(InvalidRequest, match="All arguments must have the same vector dimensions"):
+            cql.execute(f"SELECT pk, similarity_{similarity_function}(v, [1.1, 1.2, 20.25, 123.7]) FROM {table}")
+        with pytest.raises(InvalidRequest, match="All arguments must have the same vector dimensions"):
+            cql.execute(f"SELECT pk, similarity_{similarity_function}([1.0, 2.0, 3.0], [1.1, 1.2]) FROM {table}")
+        with pytest.raises(InvalidRequest, match="All arguments must have the same vector dimensions"):
+            cql.execute(f"SELECT pk, similarity_{similarity_function}([1.0, 2.0, 3.0], [1.1, 1.2, 20.25, 123.7]) FROM {table}")
```

Comment:
The one case you didn't check here (I think) is if you have different columns like
```
v1 vector<float, 3>, v2 vector<float, 3>, v3 vector<float,4>
```
It should be possible to compute the similarity of v1 and v2 (right?) but not v1 and v3.